File: ui-tests/PageObjects/Panes/CallTrace/CallTracePane.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using Microsoft.Playwright;
using UiTests.PageObjects;
using UiTests.PageObjects.Components;
using UiTests.Utils;

namespace UiTests.PageObjects.Panes.CallTrace;

/// <summary>
/// Pane representing the call trace view and its rendered call hierarchy.
/// </summary>
public class CallTracePane : TabObject
{
    private readonly ContextMenu _contextMenu;
    private List<CallTraceEntry> _entries = new();

    public CallTracePane(IPage page, ILocator root, string tabButtonText)
        : base(page, root, tabButtonText)
    {
        _contextMenu = new ContextMenu(page);
    }

    /// <summary>
    /// Locator referencing the scrolling container of call trace lines.
    /// </summary>
    public ILocator LinesContainer()
        => Root.Locator(".calltrace-lines");

    /// <summary>
    /// Locator for the search input.
    /// </summary>
    public ILocator SearchInput()
        => Root.Locator(".calltrace-search-input");

    /// <summary>
    /// Locator for the search results popup.
    /// </summary>
    public ILocator SearchResultsContainer()
        => Root.Locator(".call-search-results");

    /// <summary>
    /// Waits until call trace entries are rendered.
    /// </summary>
    public Task WaitForReadyAsync()
        => RetryHelpers.RetryAsync(async () =>
            await LinesContainer().Locator(".calltrace-call-line").CountAsync() > 0);

    /// <summary>
    /// Retrieves all currently visible call entries.
    /// </summary>
    public async Task<IReadOnlyList<CallTraceEntry>> EntriesAsync(bool forceReload = false)
    {
        if (forceReload || _entries.Count == 0)
        {
            await WaitForReadyAsync();
            var roots = await LinesContainer().Locator(".calltrace-call-line").AllAsync();
            _entries = roots
                .Select(locator => new CallTraceEntry(this, locator, _contextMenu))
                .ToList();
        }

        return _entries;
    }

    /// <summary>
    /// Clears the cached entry list so the next access reloads the DOM.
    /// </summary>
    public void InvalidateEntries() => _entries.Clear();

    /// <summary>
    /// Finds the first call trace entry whose function name matches <paramref name="functionName"/>.
    /// </summary>
    public async Task<CallTraceEntry?> FindEntryAsync(string functionName, bool forceReload = false)
    {
        var entries = await EntriesAsync(forceReload);
        foreach (var entry in entries)
        {
            var name = await entry.FunctionNameAsync();
            if (string.Equals(name, functionName, StringComparison.OrdinalIgnoreCase))
            {
                return entry;
            }
        }

        return null;
    }

    /// <summary>
    /// Performs a search query within the call trace pane.
    /// </summary>
    public async Task SearchAsync(string query)
    {
        await SearchInput().FillAsync(query);
        await SearchInput().PressAsync("Enter");
        await RetryHelpers.RetryAsync(async () =>
            await SearchResultsContainer().Locator(".search-result").CountAsync() > 0);
    }

    /// <summary>
    /// Clears the search field.
    /// </summary>
    public async Task ClearSearchAsync()
    {
        await SearchInput().FillAsync(string.Empty);
        await RetryHelpers.RetryAsync(async () =>
            await SearchResultsContainer().Locator(".search-result").CountAsync() == 0);
    }

    /// <summary>
    /// Returns the tooltip value currently rendered (if any).
    /// </summary>
    public async Task<ValueComponentView?> ActiveTooltipAsync()
    {
        var tooltip = Root.Locator(".call-tooltip");
        if (await tooltip.CountAsync() == 0)
        {
            return null;
        }

        return new ValueComponentView(tooltip.Locator(".value-expanded").First);
    }
}

File: tsc-ui-tests/page-objects/panes/call-trace/call-trace-pane.ts
import type { Locator, Page } from "@playwright/test";
import { ContextMenu } from "../../components/context-menu";
import { ValueComponentView } from "../../components/value-component-view";
import { CallTraceEntry } from "./call-trace-entry";
import { retry } from "../../../lib/retry-helpers";

/**
 * Pane representing the call trace view and its rendered call hierarchy.
 *
 * Port of ui-tests/PageObjects/Panes/CallTrace/CallTracePane.cs
 */
export class CallTracePane {
  readonly page: Page;
  readonly root: Locator;
  readonly tabButtonText: string;
  private readonly contextMenu: ContextMenu;
  private entries: CallTraceEntry[] = [];

  constructor(page: Page, root: Locator, tabButtonText: string) {
    this.page = page;
    this.root = root;
    this.tabButtonText = tabButtonText;
    this.contextMenu = new ContextMenu(page);
  }

  tabButton(): Locator {
    return this.page.locator(".lm_title", { hasText: this.tabButtonText }).first();
  }

  linesContainer(): Locator {
    return this.root.locator(".calltrace-lines");
  }

  searchInput(): Locator {
    return this.root.locator(".calltrace-search-input");
  }

  searchResultsContainer(): Locator {
    return this.root.locator(".call-search-results");
  }

  async waitForReady(): Promise<void> {
    await retry(
      async () =>
        (await this.linesContainer().locator(".calltrace-call-line").count()) > 0,
    );
  }

  async getEntries(forceReload = false): Promise<CallTraceEntry[]> {
    if (forceReload || this.entries.length === 0) {
      await this.waitForReady();
      const roots = await this.linesContainer()
        .locator(".calltrace-call-line")
        .all();
      this.entries = roots.map(
        (l) => new CallTraceEntry(this, l, this.contextMenu),
      );
    }
    return this.entries;
  }

  invalidateEntries(): void {
    this.entries = [];
  }

  /**
   * Finds the first call trace entry matching the function name.
   * Entries scrolled out of the virtualized viewport are silently skipped.
   */
  async findEntry(
    functionName: string,
    forceReload = false,
  ): Promise<CallTraceEntry | null> {
    const entries = await this.getEntries(forceReload);
    for (const entry of entries) {
      try {
        const name = await entry.functionName();
        if (name.toLowerCase() === functionName.toLowerCase()) {
          return entry;
        }
      } catch {
        // Entry may be scrolled out of the virtualized viewport
      }
    }
    return null;
  }

  async search(query: string): Promise<void> {
    await this.searchInput().fill(query);
    await this.searchInput().press("Enter");
    await retry(
      async () =>
        (await this.searchResultsContainer().locator(".search-result").count()) > 0,
    );
  }

  async clearSearch(): Promise<void> {
    await this.searchInput().fill("");
    await retry(
      async () =>
        (await this.searchResultsContainer().locator(".search-result").count()) === 0,
    );
  }

  async activeTooltip(): Promise<ValueComponentView | null> {
    const tooltip = this.root.locator(".call-tooltip");
    if ((await tooltip.count()) === 0) return null;
    return new ValueComponentView(tooltip.locator(".value-expanded").first());
  }
}
